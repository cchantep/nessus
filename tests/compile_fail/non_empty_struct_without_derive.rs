// This test verifies that a struct WITHOUT FieldWitnesses does NOT implement NonEmptyStruct

use nessus::NonEmptyStruct;
use serde::{Serialize, Deserialize};

// This struct does NOT derive FieldWitnesses
#[derive(Debug, Clone, Serialize, Deserialize)]
struct RegularStruct {
    Name: String,
    Age: i32,
}

fn main() {
    // This should fail to compile because RegularStruct does not derive FieldWitnesses
    // and therefore does not implement NonEmptyStruct
    fn _assert_implements_non_empty_struct<T: NonEmptyStruct>() {}
    _assert_implements_non_empty_struct::<RegularStruct>();
}
