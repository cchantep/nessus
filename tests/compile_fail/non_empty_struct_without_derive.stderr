error[E0277]: the trait bound `RegularStruct: NonEmptyStruct` is not satisfied
  --> tests/compile_fail/non_empty_struct_without_derive.rs:17:43
   |
17 |     _assert_implements_non_empty_struct::<RegularStruct>();
   |                                           ^^^^^^^^^^^^^ the trait `NonEmptyStruct` is not implemented for `RegularStruct`
   |
note: required by a bound in `_assert_implements_non_empty_struct`
  --> tests/compile_fail/non_empty_struct_without_derive.rs:16:47
   |
16 |     fn _assert_implements_non_empty_struct<T: NonEmptyStruct>() {}
   |                                               ^^^^^^^^^^^^^^ required by this bound in `_assert_implements_non_empty_struct`
